File: balboa-http/src/main/scala/com/socrata/balboa/server/rest/MetricsRest.scala
package com.socrata.balboa.server.rest

import java.nio.charset.StandardCharsets.UTF_8
import java.util.concurrent.TimeUnit

import com.socrata.balboa.metrics.Metric.RecordType
import com.socrata.balboa.metrics.{Metric, Metrics}
import com.socrata.balboa.metrics.data.{DataStoreFactory, DateRange, Period}
import com.socrata.balboa.metrics.impl.ProtocolBuffersMetrics
import com.socrata.balboa.server.{EntityJSON, ResponseWithType, ServiceUtils}
import com.socrata.balboa.server.ResponseWithType._
import org.codehaus.jackson.map.annotate.JsonSerialize
import org.codehaus.jackson.map.{ObjectMapper, SerializationConfig}
import org.scalatra.{Ok, NoContent, Params}

class MetricsRest

object MetricsRest {
  val dataStore = DataStoreFactory.get()

  val StartKey = "start"
  val EndKey = "end"
  val PeriodKey = "period"
  val DateKey = "date"
  val CombineKey = "combine"
  val FieldKey = "field"

  val seriesMeter = com.yammer.metrics.Metrics.newTimer(
    classOf[MetricsRest], "series queries", TimeUnit.MILLISECONDS, TimeUnit.SECONDS)

  val rangeMeter = com.yammer.metrics.Metrics.newTimer(
    classOf[MetricsRest], "range queries", TimeUnit.MILLISECONDS, TimeUnit.SECONDS)

  val periodMeter = com.yammer.metrics.Metrics.newTimer(
    classOf[MetricsRest], "period queries", TimeUnit.MILLISECONDS, TimeUnit.SECONDS)


  def get(entityId: String, params: Params, accepts: Seq[String]): ResponseWithType = {
    val period = params.get(PeriodKey).map(Extractable[Period].extract) match {
      case Some(Right(value)) => value
      case Some(Left(err)) => return badRequest(PeriodKey, err)
      case None => return required(PeriodKey)
    }

    val date = params.getOrElse(DateKey, { return required(DateKey) })
    val combine = params.get(CombineKey)
    val field = params.get(FieldKey)

    val mediaType = bestMediaType(accepts, json, protobuf).getOrElse { return unacceptable }

    val begin = System.currentTimeMillis()

    val range = DateRange.create(period, ServiceUtils.parseDate(date).getOrElse { return malformedDate(date) })

    try
    {
      val iter = dataStore.find(entityId, period, range.start, range.end)
      var metrics = Metrics.summarize(iter)

      combine.foreach { c => metrics = metrics.combine(c) }
      field.foreach { f => metrics = metrics.filter(f) }

      render(mediaType, metrics)
    }
    finally
    {
      periodMeter.update(System.currentTimeMillis() - begin, TimeUnit.MILLISECONDS)
    }
  }

  def post(entityId: String, entityOpt: Option[EntityJSON]): ResponseWithType = {
    val entity = entityOpt.getOrElse({ return required("entity") })
    val metrics = new Metrics()
    for ((name, metric) <- entity.metrics) {
      val recordType = metric.`type` match {
        case "ABSOLUTE" => RecordType.ABSOLUTE
        case "AGGREGATE" => RecordType.AGGREGATE
        case _ => return badRequest("metric type", "must be ABSOLUTE or AGGREGATE")
      }
      metrics.put(name, new Metric(recordType, metric.value))
    }
    dataStore.persist(entityId, entity.timestamp, metrics)
    ResponseWithType(json, NoContent())
  }

  def range(entityId: String, params: Params, accepts: Seq[String]): ResponseWithType = {
    val start = params.getOrElse((StartKey), { return required(StartKey) })
    val end = params.getOrElse((EndKey), { return required(EndKey) })
    val combine = params.get(CombineKey)
    val field = params.get(FieldKey)

    val startDate = ServiceUtils.parseDate(start).getOrElse { return malformedDate(start) }
    val endDate = ServiceUtils.parseDate(end).getOrElse { return malformedDate(end) }

    val mediaType = bestMediaType(accepts, json, protobuf).getOrElse { return unacceptable }

    val begin = System.currentTimeMillis()

    try
    {
      val iter = dataStore.find(entityId, startDate, endDate)
      var metrics = Metrics.summarize(iter)

      combine.foreach { c => metrics = metrics.combine(c) }
      field.foreach { f => metrics = metrics.filter(f) }

      render(mediaType, metrics)
    }
    finally
    {
      rangeMeter.update(System.currentTimeMillis() - begin, TimeUnit.MILLISECONDS)
    }
  }

  def series(entityId: String, params: Params, accepts: Seq[String]): ResponseWithType = {
    val period = params.get(PeriodKey).map(Extractable[Period].extract) match {
      case Some(Right(value)) => value
      case Some(Left(err)) => return badRequest(PeriodKey, err)
      case None => return required(PeriodKey)
    }
    val start = params.getOrElse(StartKey, { return required(StartKey) })
    val end = params.getOrElse(EndKey, { return required(EndKey) })

    val startDate = ServiceUtils.parseDate(start).getOrElse { return malformedDate(start) }
    val endDate = ServiceUtils.parseDate(end).getOrElse { return malformedDate(end) }

    bestMediaType(accepts, json).getOrElse { return unacceptable }

    val begin = System.currentTimeMillis()

    try
    {
      val body = renderJson(dataStore.slices(entityId, period, startDate, endDate)).getBytes(UTF_8)
      ResponseWithType(json, Ok(body))
    }
    finally
    {
      seriesMeter.update(System.currentTimeMillis() - begin, TimeUnit.MILLISECONDS)
    }
  }

  def bestMediaType(accepts: Seq[String], types: String*): Option[String] = {
    if(accepts.isEmpty) return Some(types(0))
    for {
      accept <- accepts
      typ <- types
    } {
      if(accept.contains(typ.replaceAll(";.*", ""))) return Some(typ)
    }
    if(accepts.exists(_.contains("*/*"))) return Some(types(0))
    None
  }

  private def render(format: String, metrics: Metrics): ResponseWithType = {
    val bytes = if(format == protobuf) {
      val mapper = new ProtocolBuffersMetrics
      mapper.merge(metrics)
      mapper.serialize
    } else {
      renderJson(metrics).getBytes(UTF_8)
    }
    ResponseWithType(format, Ok(bytes))
  }

  private def renderJson(obj: Any) =
  {
    val mapper = new ObjectMapper()
    mapper.configure(SerializationConfig.Feature.INDENT_OUTPUT, true)
    mapper.getSerializationConfig.withSerializationInclusion(JsonSerialize.Inclusion.NON_NULL)
    mapper.configure(SerializationConfig.Feature.WRITE_ENUMS_USING_TO_STRING, true)

    mapper.writeValueAsString(obj)
  }
}

File: balboa-http/src/main/java/com/socrata/balboa/server/rest/MetricsRest.java
package com.socrata.balboa.server.rest;

import com.socrata.balboa.metrics.Metrics;
import com.socrata.balboa.metrics.Timeslice;
import com.socrata.balboa.metrics.data.*;
import com.socrata.balboa.metrics.impl.ProtocolBuffersMetrics;
import com.socrata.balboa.server.ServiceUtils;
import com.socrata.balboa.server.exceptions.InvalidRequestException;
import com.yammer.metrics.core.TimerMetric;
import org.codehaus.jackson.map.ObjectMapper;
import org.codehaus.jackson.map.SerializationConfig;
import org.codehaus.jackson.map.annotate.JsonSerialize;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.io.IOException;
import java.util.Date;
import java.util.Iterator;
import java.util.concurrent.TimeUnit;

@Path("/metrics/{entityId}")
public class MetricsRest
{
    public static final TimerMetric seriesMeter = com.yammer.metrics.Metrics.newTimer(MetricsRest.class, "series queries", TimeUnit.MILLISECONDS, TimeUnit.SECONDS);
    public static final TimerMetric rangeMeter = com.yammer.metrics.Metrics.newTimer(MetricsRest.class, "range queries", TimeUnit.MILLISECONDS, TimeUnit.SECONDS);
    public static final TimerMetric periodMeter = com.yammer.metrics.Metrics.newTimer(MetricsRest.class, "period queries", TimeUnit.MILLISECONDS, TimeUnit.SECONDS);
    private static Log log = LogFactory.getLog(MetricsRest.class);

    @GET
    @Produces({"application/json", "application/x-protobuf"})
    public Response get(
            @PathParam("entityId") String entityId,
            @QueryParam("period") Period period,
            @QueryParam("date") String date,
            @QueryParam("combine") String combine,
            @QueryParam("field") String field,
            @Context HttpHeaders headers
    ) throws IOException
    {
        long begin = System.currentTimeMillis();

        try
        {
            DataStore ds = DataStoreFactory.get();
            DateRange range = DateRange.create(period, ServiceUtils.parseDate(date));

            Iterator<Metrics> iter = ds.find(entityId, period, range.start, range.end);
            Metrics metrics = Metrics.summarize(iter);

            if (combine != null)
            {
                metrics = metrics.combine(combine);
            }

            if (field != null)
            {
                metrics = metrics.filter(field);
            }

            return render(getMediaType(headers), metrics);
        }
        finally
        {
            periodMeter.update(System.currentTimeMillis() - begin, TimeUnit.MILLISECONDS);
        }
    }

    @GET
    @Path("range")
    @Produces({"application/json", "application/x-protobuf"})
    public Response range(
            @PathParam("entityId") String entityId,
            @QueryParam("start") String start,
            @QueryParam("end") String end,
            @QueryParam("combine") String combine,
            @QueryParam("field") String field,
            @Context HttpHeaders headers
    ) throws IOException
    {
        long begin = System.currentTimeMillis();

        try
        {
            DataStore ds = DataStoreFactory.get();

            Date startDate = ServiceUtils.parseDate(start);
            Date endDate = ServiceUtils.parseDate(end);

            Iterator<Metrics> iter = ds.find(entityId, startDate, endDate);
            Metrics metrics = Metrics.summarize(iter);

            if (combine != null)
            {
                metrics = metrics.combine(combine);
            }

            if (field != null)
            {
                metrics = metrics.filter(field);
            }

            return render(getMediaType(headers), metrics);
        }
        finally
        {
            rangeMeter.update(System.currentTimeMillis() - begin, TimeUnit.MILLISECONDS);
        }
    }

    @GET
    @Path("series")
    @Produces({"application/json", "application/x-protobuf"})
    public Response series(
            @PathParam("entityId") String entityId,
            @QueryParam("period") Period period,
            @QueryParam("start") String start,
            @QueryParam("end") String end,
            @Context HttpHeaders headers
    ) throws IOException
    {
        long begin = System.currentTimeMillis();

        try
        {
            DataStore ds = DataStoreFactory.get();

            Date startDate = ServiceUtils.parseDate(start);
            Date endDate = ServiceUtils.parseDate(end);

            Iterator<? extends Timeslice> iter = ds.slices(entityId, period, startDate, endDate);

            return render(getMediaType(headers), iter);
        }
        finally
        {
            seriesMeter.update(System.currentTimeMillis() - begin, TimeUnit.MILLISECONDS);
        }
    }

    @GET
    @Path("meta")
    @Produces("application/json")
    public Response meta(
            @PathParam("entityId") String entityId,
            @Context HttpHeaders headers
    ) throws IOException
    {
        throw new UnsupportedOperationException("Getting metadata is no longer implemented. Contact your System Administrator");
    }

    MediaType getMediaType(HttpHeaders headers)
    {
        MediaType format = headers.getMediaType();
        if (format == null)
        {
            for (MediaType type : headers.getAcceptableMediaTypes())
            {
                format = type;
                break;
            }
        }

        return format;
    }

    private Response render(MediaType format, Iterator<? extends Timeslice> metrics) throws IOException
    {
        if (format.getSubtype().equals("x-protobuf"))
        {
            throw new InvalidRequestException("Unable to serialize timeslices to protobuf");
        }
        else
        {
            return renderJson(metrics);
        }
    }

    private Response render(MediaType format, Metrics metrics) throws IOException
    {
        if (format.getSubtype().equals("x-protobuf"))
        {
            ProtocolBuffersMetrics mapper = new ProtocolBuffersMetrics();
            mapper.merge(metrics);

            return Response.ok(mapper.serialize(), "application/x-protobuf").build();
        }
        else
        {
            return renderJson(metrics);
        }
    }

    private Response renderJson(Object object) throws IOException
    {
        ObjectMapper mapper = new ObjectMapper();
        mapper.configure(SerializationConfig.Feature.INDENT_OUTPUT, true);
        mapper.getSerializationConfig().setSerializationInclusion(JsonSerialize.Inclusion.NON_NULL);
        mapper.configure(SerializationConfig.Feature.WRITE_ENUMS_USING_TO_STRING, true);

        return Response.ok(mapper.writeValueAsString(object), "application/json").build();
    }
}
